File: hasCycleII.py
class ListNode:
	def __init__(self, x=None):
		self.val=x 
		self.next=None

	def constructLinkedList(self, keys):
		for k in keys:
			ListNode(k)

class Solution:
	def detectCycle(self, head): # TC O(n) // SC O(n)
		seen=set()
		node=head 
		while node:
			if node in seen:
				return node 
			seen.add(node)
			node=node.next 
		return None 

	def detectCycle1(self, head): # TC O(n) // SC O(1)
		if not head: return None
		def getIntersect():
			tortoise,hare=head,head 
			while hare and hare.next:
				tortoise=tortoise.next 
				hare=hare.next.next 
				if tortoise==hare:
					return tortoise 
			return None
		intersect=getIntersect()
		if not intersect: return None 
		ptr1=head 
		ptr2=intersect 
		while ptr1!=ptr2:
			ptr1=ptr1.next 
			ptr2=ptr2.next 
		return ptr1 

head=ListNode(3)
head.next=ListNode(2)
head.next.next=ListNode(0)
head.next.next.next=ListNode(-4)
head.next.next.next=head.next 
#ll=ListNode()
#LinkedList=ll.constructLinkedList([3,2,0,-4])
s=Solution()
print(s.detectCycle(head).val)
print(s.detectCycle1(head).val)

File: hasCycleII.java
import java.util.*;
class Solution{ // TC O(n) SC O(n)
    public Node detectCycle(Node head) {
        Set<Node> visited = new HashSet<Node>();
        Node node = head;
        while (node != null) {
            if (visited.contains(node)) {
                return node;
            }
            visited.add(node);
            node = node.next;
        }
        return null;
    }
}

class Solution1{ // TC O(n) SC O(1)
    private Node getIntersect(Node head) {
        Node tortoise = head;
        Node hare = head;

        // A fast pointer will either loop around a cycle and meet the slow
        // pointer or reach the `null` at the end of a non-cyclic list.
        while (hare != null && hare.next != null) {
            tortoise = tortoise.next;
            hare = hare.next.next;
            if (tortoise == hare) {
                return tortoise;
            }
        }

        return null;
}

    public Node detectCycle(Node head) {
        if (head == null) {
            return null;
        }

        // If there is a cycle, the fast/slow pointers will intersect at some
        // node. Otherwise, there is no cycle, so we cannot find an entrance to
        // a cycle.
        Node intersect = getIntersect(head);
        if (intersect == null) {
            return null;
        }

        // To find the entrance to the cycle, we have two pointers traverse at
        // the same speed -- one from the front of the list, and the other from
        // the point of intersection.
        Node ptr1 = head;
        Node ptr2 = intersect;
        while (ptr1 != ptr2) {
            ptr1 = ptr1.next;
            ptr2 = ptr2.next;
        }

        return ptr1;
    }
}

class hasCycleII{
	public static void main(String[] args) {
		LinkedList ll=new LinkedList();
		int[] keys={3,2,0,-4};
		LinkedList list=ll.constructLinkedList(keys);
		ll.printListHead(list.head);
		// connectiong -4 to 2 
		Node node=list.head;
		while (node.next!=null){
			node=node.next;
		}
		node.next=list.head.next;
		Solution s=new Solution();
		System.out.println(s.detectCycle(list.head).data);
		Solution1 s1=new Solution1();
		System.out.println(s1.detectCycle(list.head).data);
		//Node result=s.detectCycle(list.head);
		//ll.printListHead(result);
	}
}